File: ljh/level2/오픈채팅방/Solution.java
/**
 * 오픈 채팅방(220205-220211)
 */

import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        Map<String, String> openChat = new HashMap<>();

        String[] splited;
        String state;
        String id;
        String nickname;
        int changeCount = 0;

        for(String e: record) {
            splited = e.split(" ");
            state = splited[0];

            if(state.equals("Leave")){
                continue;
            }

            id = splited[1];
            nickname = splited[2];

            if(state.equals("Change")){
                changeCount++;
                openChat.replace(id, nickname);
                continue;
            }

            if(openChat.containsKey(splited[1])){
                openChat.replace(id, nickname);
            }else{
                openChat.put(id, nickname);
            }

        }

        String[] result = new String[record.length-changeCount];
        int index = 0;
        StringBuffer sb = new StringBuffer();

        for(String e: record) {
            splited = e.split(" ");
            state = splited[0];
            id = splited[1];
            nickname = openChat.get(id);

            sb.append(nickname);
            sb.append("님이 ");

            switch(state) {
                case "Enter":
                    sb.append("들어왔습니다.");
                    result[index++] = sb.toString();
                    break;
                case "Leave":
                    sb.append("나갔습니다.");
                    result[index++] = sb.toString();
                    break;
            }

            sb.setLength(0);
        }

        return result;
    }

}

File: ljh/level2/오픈채팅방/Solution.js

// 오픈 채팅방(220205-220211)

function solution(record) {
    const id_map = new Map();
    const result = [];

    for(let m of record) {
        let message = m.split(' ');

        if(message[0] !== 'Leave') {
            id_map.set(message[1], message[2]);
        }
    }

    for(let m of record) {
        let message = m.split(' ');

        switch(message[0]) {
            case 'Enter':
                result.push(id_map.get(message[1]) + '님이 들어왔습니다.');
                break;
            case 'Leave':
                result.push(id_map.get(message[1]) + '님이 나갔습니다.');
                break;
        }
    }

    return result;
}